src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -0,0 +1,396 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    internal partial class DiagnosticIncrementalAnalyzer
+    {
+        /// <summary>
+        /// State for diagnostics that belong to a project at given time.
+        /// </summary>
+        private class ProjectState
+        {
+            // project id of this state
+            private readonly StateSet _owner;
+
+            // last aggregated analysis result for this project saved
+            private AnalysisResult _lastResult;
+
+            public ProjectState(StateSet owner, ProjectId projectId)
+            {
+                _owner = owner;
+                _lastResult = new AnalysisResult(projectId, VersionStamp.Default, documentIds: null, isEmpty: true);
+            }
+
+            public ImmutableHashSet<DocumentId> GetDocumentsWithDiagnostics()
+            {
+                return _lastResult.DocumentIdsOrEmpty;
+            }
+
+            public bool IsEmpty()
+            {
+                return _lastResult.IsEmpty;
+            }
+
+            public bool IsEmpty(DocumentId documentId)
+            {
+                return IsEmpty(_lastResult, documentId);
+            }
+
+            public async Task<AnalysisResult> GetAnalysisDataAsync(Project project, bool avoidLoadingData, CancellationToken cancellationToken)
```

Comment:
Add a comment here on how this differs from GetProjectAnalysisDataAsync which takes the same parameters?